core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableTopNHeapSort.java

```diff
@@ -0,0 +1,162 @@
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to you under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.calcite.adapter.enumerable;
+
+import org.apache.calcite.DataContext;
+import org.apache.calcite.linq4j.tree.BlockBuilder;
+import org.apache.calcite.linq4j.tree.Expression;
+import org.apache.calcite.linq4j.tree.Expressions;
+import org.apache.calcite.plan.RelOptCluster;
+import org.apache.calcite.plan.RelOptCost;
+import org.apache.calcite.plan.RelOptPlanner;
+import org.apache.calcite.plan.RelTraitSet;
+import org.apache.calcite.rel.RelCollation;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.core.Sort;
+import org.apache.calcite.rel.metadata.RelMetadataQuery;
+import org.apache.calcite.rex.RexDynamicParam;
+import org.apache.calcite.rex.RexLiteral;
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.util.BuiltInMethod;
+import org.apache.calcite.util.Pair;
+import org.apache.calcite.util.Util;
+
+/**
+ * Implementation of {@link org.apache.calcite.rel.core.Sort} in
+ * {@link org.apache.calcite.adapter.enumerable.EnumerableConvention enumerable calling convention}.
+ * It also deals with the offset and the limit.
+ */
+public class EnumerableTopNHeapSort extends Sort implements EnumerableRel {
+
+  /**
+   * Creates an EnumerableTopNHeapSort.
+   *
+   * <p>Use {@link #create} unless you know what you're doing.
+   */
+  public EnumerableTopNHeapSort(
+      RelOptCluster cluster,
+      RelTraitSet traitSet,
+      RelNode input,
+      RelCollation collation,
+      RexNode offset,
+      RexNode fetch) {
+    super(cluster, traitSet, input, collation, offset, fetch);
+    assert this.getConvention() instanceof EnumerableConvention;
+    assert this.getConvention() == input.getConvention();
+    assert fetch != null;
+  }
+
+  /** Creates an EnumerableTopNHeapSort. */
+  public static EnumerableTopNHeapSort create(
+      RelNode child,
+      RelCollation collation,
+      RexNode offset,
+      RexNode fetch) {
+    final RelOptCluster cluster = child.getCluster();
+    final RelTraitSet traitSet = cluster.traitSetOf(EnumerableConvention.INSTANCE).replace(
+        collation);
+    return new EnumerableTopNHeapSort(cluster, traitSet, child, collation, offset, fetch);
+  }
+
+  private static Expression getExpression(RexNode offset) {
```

Comment:
minor: parameter name should be more general (e.g. simply `rexNode`). In practice it can be not just the `offset`, but also the `fetch` that is used we this method is called.

Reply:
I just meant that the method's signature could look like `private static Expression getExpression(RexNode rexNode)`. Anyway, as @julianhyde said, probably this method is not really necessary because we can assume that fetch / offset are always RexLiterals, so probably we could remove this method all together.

Reply:
I was mistaken. Dynamic parameters are supported in `EnumerableLimit`, so they should be in this, too.

Can you share the code with `EnumerableLimit`?

Add a test that uses dynamic parameters.